code/modules/xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,217 @@
+/obj/item/xenoartifact_labeler
+    name = "Xenoartifact Labeler"
+    icon = 'icons/obj/xenoarchaeology/xenoartifact_tech.dmi'
+    icon_state = "xenoartifact_labeler"
+    desc = "A tool scientists use to label their alien bombs."
+    throw_speed = 3
+    throw_range = 5
+    w_class = WEIGHT_CLASS_TINY
+
+    var/list/activator = list(null) //Checked trait
+    var/list/activator_traits = list() //Display names
+
+    var/list/minor_trait = list(null, null, null)
+    var/list/minor_traits = list()
+
+    var/list/major_trait = list(null)
+    var/list/major_traits = list()
+
+    var/list/malfunction = list(null)
+    var/list/malfunction_list = list()  
+
+    var/list/info_list = list() //trait dialogue essentially
+
+    var/sticker_name //Name artifacts something pretty
+    var/list/sticker_traits = list()//passed down to sticker
+
+/obj/item/xenoartifact_labeler/Initialize()
+    . = ..()
+    get_trait_list_desc(activator_traits, /datum/xenoartifact_trait/activator) //I forgot why this is alone here. Once again, I'd rather not change shit now.
+
+/obj/item/xenoartifact_labeler/interact(mob/user)
+    //ui_interact(user, "XenoartifactLabeler")
+    ..()
+
+/obj/item/xenoartifact_labeler/ui_interact(mob/user, datum/tgui/ui)
+    ui = SStgui.try_update_ui(user, src, ui)
+    if(!ui)
+        ui = new(user, src, "XenoartifactLabeler")
+        ui.open()
+
+/obj/item/xenoartifact_labeler/ui_data(mob/user)
+    var/list/data = list()
+    data["activator"] = activator
+    data["activator_traits"] = get_trait_list_desc(activator_traits, /datum/xenoartifact_trait/activator)
+
+    data["minor_trait"] = minor_trait
+    data["minor_traits"] = get_trait_list_desc(minor_traits, /datum/xenoartifact_trait/minor)
+
+    data["major_trait"] = major_trait
+    data["major_traits"] = get_trait_list_desc(major_traits, /datum/xenoartifact_trait/major)
+
+    data["malfunction"] = malfunction
+    data["malfunction_list"] = get_trait_list_desc(malfunction_list, /datum/xenoartifact_trait/malfunction)
+
+    data["info_list"] = info_list
+
+    return data
+
+/obj/item/xenoartifact_labeler/ui_act(action, params)
+    if(..())
+        return
+
+    if(action == "print_traits")
+        create_label(sticker_name)
+
+    if(action == "change_print_name")
+        sticker_name = params["name"]
```

Comment:
❗ Href exploit: Params["name"] is not sanatised and could be any value, allowing you to introduce anything into the name of a sticker.

Reply:
ToDo premerge.
I'll have to research this one.